src/AWS4AuthRequest.jl

```diff
@@ -0,0 +1,156 @@
+module AWS4AuthRequest
+
+using Base64
+using Dates
+using Unicode
+using MbedTLS: digest, MD_SHA256, MD_MD5
+import ..Layer, ..request, ..Headers
+using ..URIs
+using ..Pairs: getkv, setkv, rmkv
+import ..@debug, ..DEBUG_LEVEL
+
+
+"""
+    request(AWS4AuthLayer, ::URI, ::Request, body) -> HTTP.Response
+
+Add a [AWS Signature Version 4](http://docs.aws.amazon.com/general/latest/gr/signature-version-4.html)
+`Authorization` header to a `Request`.
+
+
+Credentials are read from environment variables `AWS_ACCESS_KEY_ID`,
+`AWS_SECRET_ACCESS_KEY` and `AWS_SESSION_TOKEN`.
```

Comment:
should also mention that they can be passed as keyword arguments like `aws_access_key_id` (if I understand the implementation below correctly)

Reply:
Why include AWS stuff in HTTP? 🤔

Reply:
@quinnj, see kw arg doc at top-level request function: https://github.com/samoconnor/HTTP.jl/blob/simplify_parser_branch/src/HTTP.jl#L143

@EricForgy, 1) because it's needed support tests that use S3; 2) Because it served as an example of testing for myself how easy it would be for a user to add a custom layer to the stack; 3) because it seems like a natural peer of BasicAuthRequest.jl.
See also comments here about eventually splitting this all up into seperate packages: https://github.com/JuliaWeb/WebSockets.jl/issues/84#issuecomment-356110858

Reply:
Yeah, this is not that much code for a pretty useful piece of functionality; it could easily move to AWSCore in the future, but for testing purposes, I'm totally ok w/ it living here.

Reply:
The code is copy-pasted from AWSCore.jl.

Moving it back there at some point will serve as an interesting example of stack customisation because of how it interacts with other layers. e.g. The auth headers need to be recomputed after a redirect, and they need to know _all_ the headers in the request, so the auth layer has to go at a particular layer in the stack. It will be nice to have an example of this for other users to follow.